### ISSUE TITLE
[false-positive] Param incorrectly marked as unused

### ISSUE THREAD

**Issue by aidantwoods** [vimeo/psalm#1980]
Apologies for not including a psalm.dev link here, I tried a few things but wasn't able to create a self-contained repro of the issue, it *does* however occur on latest master (at time of reporting: 7ded79bb2). I figured it might be best to just file this anyway rather than leaving  the issue unreported—I'll keep trying on finding a minimal PoC and update this if I track down what causes this.

The issue is reproducible locally and [in travis](https://travis-ci.org/erusev/parsedown/jobs/563657541#L569-L581), and is as follows:

```
$ vendor/bin/psalm --find-dead-code

ERROR: PossiblyUnusedParam - src/Parsedown.php:84:55 - Param #2 is never referenced in this method
    public static function blocks(Lines $Lines, State $State)


------------------------------
1 errors found
------------------------------

Checks took 1.12 seconds and used 153.612MB of memory
Psalm was able to infer types for 99.7719% of the codebase
Script vendor/bin/psalm --find-dead-code handling the test-dead-code event returned with error code 1
```

and the method in question looks like:

```php
    /**
     * @return array{Block[], State}
     */
    public static function blocks(Lines $Lines, State $State)
    {
        $RecursionLimiter = $State->get(RecursionLimiter::class)->increment();

        if ($RecursionLimiter->isDepthExceeded()) {
            $State = $State->setting(new BlockTypes([], []));
        }

        $State = $State->setting($RecursionLimiter);

        // ...
```

where clearly, param 2 (`$State`) has in-fact been used.

You can obtain a copy of the affected codebase by running:

```bash
git clone https://github.com/erusev/parsedown \
  && cd parsedown \
  && git checkout 18ee4ff \
  && composer install
```

and reproduce the issue with:

```bash
vendor/bin/psalm --find-dead-code
```

**Comment by muglug** [comment]
Hmm. This is definitely a bug (relating, I think, to the `continue 2` statement) but absent a more concise reproducer there's not much I can do at the moment.

**Comment by aidantwoods** [comment]
Well you were definitely correct about that: I tried dropping the `continue 2` statement and that removed `PossiblyUnusedParam` error. I tried also dropping lines that mutated `$State` and found that removing [line 134](https://github.com/erusev/parsedown/blob/18ee4ffe04f4b07aa38dd3ba36db2a1878b40887/src/Parsedown.php#L134) `$State = $Block->latestState();` also "fixes" the error (this is present in the same loop as the `continue 2;` statement so possibly has something to do with it).

I managed to repro a *similar* error, but not quite the same one—they *might* have the same cause 🤷‍♂

https://psalm.dev/r/42cf94dc3d

**Comment by muglug** [comment]
Thanks for your effort! Further narrowed down to

```php
class Foo {}

function takesFoo(Foo $Foo1, Foo $Foo2): void {
    while (rand(0, 1)) {
        echo get_class($Foo1);

        if (rand(0, 1)) {
            $Foo1 = $Foo2;

            continue;
        }
    }

    echo get_class($Foo1);
}
```

https://psalm.dev/r/157d559425

**Comment by aidantwoods** [comment]
Wooo thank you! Can confirm this fixed the `PossiblyUnusedParam` issue I was having along with these repros 😃
